### ISSUE TITLE
better gun depression/elevation for several units

### ISSUE THREAD

**Issue by shibumi** [spring1944/spring1944#323]
Hi,
Currently all tanks have full gun depression/elevation (according to mario). Other units like Ho-Ro that use indirect fire has no full gun depression/elevation. This leads to the bug that ho-ro can't shoot down from hills although it should be able to due to indirect fire. Maybe we should give all units with indirect fire full depression/elevation, that it can finally hit targets that lies lower or upper when the unit. Same for Rocket mortar and most Howitzer btw..

**Comment by sanguinariojoe** [comment]
+1

**Comment by shibumi** [comment]
Note: this would also make it finally possible to plant AT guns on a higher position like hills and shoot down the hills (more tactic). I am thinking especially about maps like Titan (the cliffs near the hills are  perfect for this.. or shooting from that hills into the middle)

**Comment by buumi** [comment]
What do you mean? You can shoot e.g. from Titan's hill to center, with all Ho-Ro, AT guns and rocket mortar. I've been doing this for years. Of course the guns need to be close to edge so that ground does not block the shot and you also can't fire right under the hill due to limited angle that weapon can shoot to. I guess you mean something else there?

**Comment by shibumi** [comment]
@buumi nice when it works for you. Last time when I tried this it doesn't work and as you said you need to move the Ho-Ro, AT guns at the edge and the edge has to have a small depression down hill otherwise it will not work. The Ho-Ro uses indirect fire so it should be able to fire in a bow and even hit targets that are near to the cliff. Same for Howitzers. And another good example where it doesn't work or it's more a case of luck is that hill in that porky map with sea and the small corridor.. I forgot the name. It's the map we talked about securing the corridor at the hill with tons of AT guns if you remember it..  Ho-Ro for example should be able to shoot down from that hill and hit all AT guns on that position, but it can't.

**Comment by simonpuchert** [comment]
If I understand correctly, the current code does a scalar product / cosine based angle check from a base direction. This does not adequately describe weapons where the range of elevation and the range of traverse differ significantly. It should be possible to change that, but it will make the code more complicated.

Then there is the problem of indirect fire. I guess that can be faked by giving an indirect-fire gun a larger elevation range (but not actually moving the barrel up/down to that pitch angle). Alternatively, one could simulate the projectile trajectory to check if a target can be reached but that would require a much more expensive computation.

Concerning the above discussion, positioning guns properly is of course an important part of the game, but we can certainly find a solution that preserves this importance while avoiding "pixel-perfect placement" shenanigans.

**Comment by sanguinariojoe** [comment]
@yuritch , please, give us some feedback

@jal76 , can you take a look?

**Comment by FLOZi** [comment]
@simonpuchert is essentially correct that current (engine) code does not differentiate between traverse and elevation. If we really wanted to do so, we would have to re-implement it lua side, which might be possible with my extended vector api but I'm not sure how fast it is.

**Comment by yuritch** [comment]
Well, a simple way is to give a large cone in the unit def (so that the elevation is covered), and then check for heading limits on the lua side. Forcing a re-targeting if the target is out of bounds. That will likely work ok with automatic attacks (fight commands, etc) and not so well with manual attack orders.

**Comment by FLOZi** [comment]
Yeah I just thought the same now, which might not even need vector code. Not sure forcing retargetting will work though, if it is the only target in range (consider SP arty) re-target won't force a re-alignment of the hull afaik.

**Comment by shibumi** [comment]
@FLOZi  what's the issue with just giving all guns gun depression? It's unrealistic that a ho-ro can depress it's gun that much, yes, but it would fix all issues without intervening with manual or automatic attacks.

**Comment by yuritch** [comment]
The issue is that the engine only provides an aiming cone, while what we are talking about is more pyramid-shaped.
So, if we provide a cone with enough angle that depression works reliably, it will also allow the gun to aim sideways, which won't look good on the same Ho-Ro

**Comment by FLOZi** [comment]
This may illustrate it for the uninitiated:

![Rhino4_2020-12-28_21-19-22](https://user-images.githubusercontent.com/302893/103243730-a340a200-4952-11eb-8940-700328c8a83d.png)

Spring gives us the red cone (essentially one dimensional - the radius), as yuri says a more realistic system for S44 is 2 dimensional, large elevation and limited traverse (the green rectangular pyramid).
